AML Wrapper for ddm_predictor.py

akzaidi opened the pull request:
## Soundtrack

- [EABS - Raise Your Hearts, Drop Your Guns](https://www.youtube.com/watch?v=Zh88rXOMw5U)

## Description

Updates the `ddm_predictor.py` script to utilize Ray for training your policy rather than Bonsai.

### Usage

* Run using Ray-on-AML with: `python ddm_predictor.py simulator.policy=ray-aml`.
* Run locally with Ray with: `python ddm_predictor.py simulator.policy=ray-local`.

Fixes # (issue): NA

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

akzaidi committed:
ADD: ray-aml integration for ddm_predictor + job.yml for aml submission with v2

akzaidi committed:
update: ray-aml integration for ddm_predictor + job.yml for aml submission with v2

akzaidi committed:
:BUG: fix: hydra parameters

akzaidi committed:
UPDATE: ddm aml integration for predictor. updated the environment yml and included the job.yml for running jobs

akzaidi commented:
@juanvergaramunoz - I'd like to make one more edit today that will make things a bit easier, which is to convert the DDM repo into a package that we can pip install. That will make things easier for `ray-on-aml` with installing the ddm scripts and ensures they are in the correct location when running the remote jobs. I still wanted to publish this so you can test this, but perhaps you can focus on the test local path before trying the longer running tests on AML?

thisisvergara commented:
@akzaidi, did you see the pytest failed? I think we need to setup numpy version to be >=1.16 & <1.20 -- and it seems currently we ask for 1.22.0

On a different note, if you are close to pushing the commit with the change you indicated above, I will hold my review. Otherwise, I will review what is already there.

thisisvergara committed:
Clarify error when model files are not available

- print error message after clarification to run `python ddm_trainer.py`

akzaidi committed:
fixup: Format Python code with Black

akzaidi committed:
UPDATE: fixes working directory from where job initiates

akzaidi committed:
UPDATE: assign IPs automatically, include a num_iters flag for ray-aml training

thisisvergara commented:
Numpy error identified that causes verbose logging to the point of storage overflow. Adding the `object` type to the creation of the numpy input data array in L462 (`X = np.array(input_array, dtype=object).reshape(1, -1)`), removed the following verbose error:
```
ddm_predictor.py:462: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
   X = np.array(input_array).reshape(1, -1)
```

Removal of that message causes an output file size reduction of x9. Reducing the original output log file from 12GB to only 1.3GB

*Additionally, removal of a print that indicated "No signal builder used" further reduced the output log file by another factor of x3.*

thisisvergara committed:
Fix log vervosity

- modified signal builder log from `print()` to `logger.info()`.
- add `object` type to creation of input data numpy array.
-- this change shrunk output log file size by an order of x9.

thisisvergara committed:
add Ray Tuner as algo training method

- modify script to use ray.tune.tuner instead of the algo object.
-- it will permit easier introduction of checkpoint saving.

thisisvergara committed:
add callback functionality to track metrics

- incorporate functionality to keep track of training reward evolution.
- use pretty_print

thisisvergara committed:
add checkpointing with predefined output folder

- add checkpoint functionality.
- define output folder to point to AML output folder.
- update Ray to 2.5 to use latest version of output folder definition.

thisisvergara committed:
add rl agent training instructions to readme

- add instructions for training your rl agent using ray rllib.

akzaidi committed:
fixup: Format Python code with Black
